src/msw/window.cpp

```diff
@@ -4705,6 +4705,89 @@ wxWindowMSW::MSWOnMeasureItem(int id, WXMEASUREITEMSTRUCT *itemStruct)
     return false;
 }
 
+// ---------------------------------------------------------------------------
+// DPI
+// ---------------------------------------------------------------------------
+
+/*extern*/
+int wxGetSystemMetrics(int nIndex, wxWindow* win)
+{
+    wxTopLevelWindow* tlw = NULL;
+    if ( win )
+    {
+        tlw = wxDynamicCast(wxGetTopLevelParent(win), wxTopLevelWindow);
+    }
+    else
+    {
+        wxWindow* window = static_cast<wxApp*>(wxApp::GetInstance())->GetTopWindow();
+        if ( window )
+        {
+            tlw = wxDynamicCast(wxGetTopLevelParent(window), wxTopLevelWindow);
+        }
+    }
+
+    if ( tlw )
+    {
+        wxDynamicLibrary dllUser32;
+        if ( dllUser32.Load(wxS("User32.dll"), wxDL_VERBATIM | wxDL_QUIET) )
+        {
+            typedef int(WINAPI *GetSystemMetricsForDpi_t)(int nIndex, UINT dpi);
+            GetSystemMetricsForDpi_t wxDL_INIT_FUNC(pfn, GetSystemMetricsForDpi, dllUser32);
+
+            if ( pfnGetSystemMetricsForDpi )
+            {
+                const int y = ::GetDeviceCaps(::GetDC(tlw->GetHWND()), LOGPIXELSY);
+                return pfnGetSystemMetricsForDpi(nIndex, (UINT)y);
+            }
+        }
+    }
+
+    return ::GetSystemMetrics(nIndex);
+}
+
+/*extern*/
+bool wxSystemParametersInfo(UINT uiAction, UINT uiParam, PVOID pvParam, UINT fWinIni, wxWindow* win)
+{
+    // Only use SystemParametersInfoForDpi for the following uiActions:
+    // https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-systemparametersinfofordpi
+    if ( !(uiAction == SPI_GETICONTITLELOGFONT || uiAction == SPI_GETICONMETRICS || uiAction == SPI_GETNONCLIENTMETRICS) )
+    {
+        return ::SystemParametersInfo(uiAction, uiParam, pvParam, fWinIni);
+    }
+
+    wxTopLevelWindow* tlw = NULL;
```

Comment:
This code should be extracted into a separate function to avoid duplicating it between this one and `wxGetSystemMetrics()` or maybe even to call it explicitly (see the global review comment).